File: [23.12]Design-Pattern/Java/src/structural/flyweight/Tree.java
package structural.flyweight;

public class Tree {
    Model model;
    double x;
    double y;

    private Tree(Model model, double x, double y) {
        this.model = model;
        this.x = x;
        this.y = y;
    }

    public static class Factory {
        public static Tree getInstance(String type) {
            Model model = Model.Factory.getInstance(type);
            double x = Math.random() * 10000;
            double y = Math.random() * 10000;

            System.out.println(type + "의 좌표: x=" + x + ", y=" + y);
            return new Tree(model, x, y);
        }
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/structural/flyweight/Tree.kt
package structural.flyweight

class Tree private constructor(
    val type: Model,
    val x: Double,
    val y: Double
) {
    companion object Factory {
        fun getInstance(type: String): Tree {
            val model = Model.Factory.getInstance(type)
            val x = Math.random() * 10000
            val y = Math.random() * 10000

            println("$type 위치: x=${x}, y=${y}")
            return Tree(model, x, y)
        }
    }
}